t/t2201-job-cmd.t

```diff
@@ -305,4 +306,278 @@ test_expect_success 'flux-job: kill fails for wrong userid' '
+	cat <<-EOT >raiseall_bs2.exp &&
+	flux-job: no states specified
+	EOT
+	test_cmp raiseall_bs2.exp raiseall_bs2.err
+'
+
+test_expect_success 'flux job: raiseall with invalid severity fails' '
+	test_must_fail flux job raiseall --severity=9 test 2>raiseall_sev.err &&
+	cat <<-EOT >raiseall_sev.exp &&
+	flux-job: raiseall: invalid exception severity
+	EOT
+	test_cmp raiseall_sev.exp raiseall_sev.err
+'
+
+test_expect_success 'flux job: raiseall with invalid type fails' '
+	test_must_fail flux job raiseall "bad type" 2>raiseall_bt.err &&
+	cat <<-EOT >raiseall_bt.exp &&
+	flux-job: raiseall: invalid exception type
+	EOT
+	test_cmp raiseall_bt.exp raiseall_bt.err
+'
+
+test_expect_success 'flux-job: raiseall --user all fails for guest' '
+	id=$(($(id -u)+1)) &&
+        test_must_fail runas ${id} \
+		flux job raiseall --user all test 2> raiseall_all_guest.err &&
+	cat <<-EOT >raiseall_all_guest.exp &&
+	flux-job: raiseall: guests can only raise exceptions on their own jobs
+	EOT
+	test_cmp raiseall_all_guest.exp raiseall_all_guest.err
+'
+
+test_expect_success 'flux-job: raiseall --user <guest_uid> works for guest' '
+	id=$(($(id -u)+1)) &&
+        runas ${id} \
+		flux job raiseall --user ${id} test 2> raiseall_guest.err
+'
+
+test_expect_success 'flux-job: raiseall --user name works' '
+	name=$(id -un) &&
+	flux job raiseall --user ${name} test 2> raiseall_name.err
+'
+
+test_expect_success 'submit 3 test jobs' '
+	for i in $(seq 1 3); do flux mini submit sleep 60 >>jobs; done
+'
+
+test_expect_success 'flux-job: raiseall returns correct count' '
+	flux job raiseall -s 7 test not_raised 2>raiseall_test.err &&
+	cat <<-EOT >raiseall_test.exp &&
+	flux-job: Command matched 3 jobs (-f to confirm)
+	EOT
+	test_cmp raiseall_test.exp raiseall_test.err
+'
+
+test_expect_success 'flux-job: raiseall -f returns correct count' '
+	flux job raiseall -s 7 -f test raised1 2>raiseall_testf.err &&
+	cat <<-EOT >raiseall_testf.exp &&
+	flux-job: Raised exception on 3 jobs (0 errors)
+	EOT
+	test_cmp raiseall_testf.exp raiseall_testf.err
+'
+
+test_expect_success 'flux-job: cancelall -f returns correct count' '
+	flux job cancelall -f 2>cancelall_testf.err &&
+	cat <<-EOT >cancelall_testf.exp &&
+	flux-job: Canceled 3 jobs (0 errors)
+	EOT
+	test_cmp cancelall_testf.exp cancelall_testf.err
+'
+
+test_expect_success 'flux job: the queue is empty' '
+	run_timeout 60 flux job drain
+'
+test_expect_success 'flux job: retrieve eventlogs' '
+	for id in $(cat jobs); do flux job eventlog ${id} >>eventlogs; done
+'
+
+test_expect_success 'flux-job: unconfirmed exception was not raised' '
+	test_must_fail grep not_raised eventlogs
+'
+
+test_expect_success 'flux-job: confirmed non-fatal exception was raised' '
+	count=$(grep raised1 eventlogs | wc -l) &&
+	test ${count} -eq 3
+'
+
+test_expect_success 'flux-job: fatal cancel exception was raised' '
+	count=$(grep cancel eventlogs | wc -l) &&
+	test ${count} -eq 3
+'
```

Comment:
When run with `FLUX_TEST_VALGRIND=t` this check is reliably failing:
```
expecting success: 
	count=$(grep cancel eventlogs | wc -l) &&
	test ${count} -eq 3

not ok 92 - flux-job: fatal cancel exception was raised
#	
#		count=$(grep cancel eventlogs | wc -l) &&
#		test ${count} -eq 3
#	

```

```console
$ cat trash-directory.t2201-job-cmd/eventlogs 
1580404899.275410 submit userid=6885 priority=16 flags=0
1580404899.355038 depend
1580404899.893830 exception type="test" severity=7 userid=6885 note="raised1"
1580404899.508602 submit userid=6885 priority=16 flags=0
1580404899.574790 depend
1580404899.893124 exception type="test" severity=7 userid=6885 note="raised1"
1580404899.944996 exception type="cancel" severity=0 userid=6885 note=""
1580404899.945371 clean
1580404899.726398 submit userid=6885 priority=16 flags=0
1580404899.794859 depend
1580404899.892191 exception type="test" severity=7 userid=6885 note="raised1"
1580404899.943833 exception type="cancel" severity=0 userid=6885 note=""
1580404899.944316 clean

```

Reply:
Excellent catch.  I was assuming in the test that eventlogs must be complete in the KVS if the queue is empty following a `flux job drain`.  However, `drain` is synchronized on the hash of active jobs, _not_ on the (batched) KVS commit for the eventlogs.

The job manager is careful not to publish state transition events until the KVS commit of the eventlog completes;  I think to avoid surprises like this, drain should also block until eventlogs are committed.